pfcpiface/conn.go

```diff
@@ -136,35 +160,52 @@ func pfcpifaceMainLoop(upf *upf, accessIP, coreIP, sourceIP, smfName string) {
 				}
 			}
 		case message.MsgTypeAssociationSetupResponse:
-			cpConnected = handleAssociationSetupResponse(msg, addr, sourceIP, accessIP)
+			cpConnected = handleAssociationSetupResponse(msg, pPkt.Address, sourceIP, accessIP)
 			if manageConnection {
 				// pass on information to go routine that result of association response
 				cpConnectionStatus <- cpConnected
 			}
 		case message.MsgTypePFDManagementRequest:
-			outgoingMessage = pconn.handlePFDMgmtRequest(upf, msg, addr, sourceIP)
+			outgoingMessage = pconn.handlePFDMgmtRequest(upf, msg, pPkt.Address, sourceIP)
 		case message.MsgTypeSessionEstablishmentRequest:
-			outgoingMessage = pconn.handleSessionEstablishmentRequest(upf, msg, addr, sourceIP)
+			outgoingMessage = pconn.handleSessionEstablishmentRequest(upf, msg, pPkt.Address, sourceIP)
 		case message.MsgTypeSessionModificationRequest:
-			outgoingMessage = pconn.handleSessionModificationRequest(upf, msg, addr, sourceIP)
+			outgoingMessage = pconn.handleSessionModificationRequest(upf, msg, pPkt.Address, sourceIP)
 		case message.MsgTypeHeartbeatRequest:
-			outgoingMessage = handleHeartbeatRequest(msg, addr)
+			outgoingMessage = handleHeartbeatRequest(msg, pPkt.Address)
 		case message.MsgTypeSessionDeletionRequest:
-			outgoingMessage = pconn.handleSessionDeletionRequest(upf, msg, addr, sourceIP)
+			outgoingMessage = pconn.handleSessionDeletionRequest(upf, msg, pPkt.Address, sourceIP)
 		case message.MsgTypeAssociationReleaseRequest:
-			outgoingMessage = handleAssociationReleaseRequest(msg, addr, sourceIP, accessIP)
+			outgoingMessage = handleAssociationReleaseRequest(msg, pPkt.Address, sourceIP, accessIP)
 			cleanupSessions()
 		default:
 			log.Println("Message type: ", msg.MessageTypeName(), " is currently not supported")
-			continue
+			return
 		}
 
 		// send the response out
 		if outgoingMessage != nil {
-			if _, err := conn.WriteTo(outgoingMessage, addr); err != nil {
+			if _, err := conn.WriteTo(outgoingMessage, pPkt.Address); err != nil {
 				log.Fatalln("Unable to transmit association setup response", err)
 			}
 		}
 
 	}
+
+	go pfcpPacketReader(pfcpRcvdPktsChan)
+
+	for {
+		select {
+
+		// create goroutine for each packet parsing
+		case rPkt := <-pfcpRcvdPktsChan:
+			go pfcpPacketParsing(rPkt, pfcpParsedPktsChan)
+
```

Comment:
depending on parsing speed of packets the order in which packets are handled might change. like If modification is handled after delete it would be difficult to identify the problem.

Reply:
Hmm...right if back to back messages come for same user and then there is chance of re-ordering. Wondering how this can be solved.